js/woo.js

```diff
@@ -0,0 +1,1469 @@
+                    '-1101': InvalidOrder, // { "code": -1101,  "message": "The risk exposure for client is too high, it may cause by sending too big order or the leverage is too low. please refer to client info to check the current exposure." }
+                    '-1102': InvalidOrder, // { "code": -1102,  "message": "The order value (price * size) is too small." }
+                    '-1103': InvalidOrder, // { "code": -1103,  "message": "The order price is not following the tick size rule for the symbol." }
+                    '-1104': InvalidOrder, // { "code": -1104,  "message": "The order quantity is not following the step size rule for the symbol." }
+                    '-1105': InvalidOrder, // { "code": -1105,  "message": "Price is X% too high or X% too low from the mid price." }
+                },
+                'broad': {
+                },
+            },
+            'precisionMode': TICK_SIZE,
+        });
+    }
+
+    async fetchMarkets (params = {}) {
+        const [ marketType, query ] = this.handleMarketTypeAndParams ('fetchMarkets', undefined, params);
+        const method = this.getSupportedMapping (marketType, {
+            'spot': 'v1PublicGetInfo',
+        });
+        let data = undefined;
+        if (marketType === 'spot') {
+            const response = await this[method] (query);
+            //
+            // {
+            //     rows: [
+            //         {
+            //             symbol: "SPOT_AAVE_USDT",
+            //             quote_min: 0,
+            //             quote_max: 100000,
+            //             quote_tick: 0.01,
+            //             base_min: 0.01,
+            //             base_max: 7284,
+            //             base_tick: 0.0001,
+            //             min_notional: 10,
+            //             price_range: 0.1,
+            //             created_time: "0",
+            //             updated_time: "1639107647.988",
+            //             is_stable: 0
+            //         },
+            //         ...
+            //     success: true
+            // }
+            //
+            data = this.safeValue (response, 'rows', []);
+        }
+        const result = [];
+        for (let i = 0; i < data.length; i++) {
+            const market = data[i];
+            const marketId = this.safeString (market, 'symbol');
+            const parts = marketId.split ('_');
+            const baseId = parts[1];
+            const quoteId = parts[2];
+            const marketTypeVal = parts[0].toLowerCase ();
```

Comment:
Let's use safeString on all keys and indices as much as possible, except in for-loops:

```JavaScript
const marketType = this.safeStringLower (parts, 0);
```